mypy/checkexpr.py

```diff
@@ -2993,3 +3203,44 @@ def map_formals_to_actuals(caller_kinds: List[int],
         for actual in actuals:
             actual_to_formal[actual].append(formal)
     return actual_to_formal
+
+
+def merge_typevars_in_callables_by_name(
```

Comment:
Maybe we can avoid adding this function here, and instead ensure that during semantic analysis there is a single binding for a type variable across overload variants. This would be a clearer solution IMO, but the current code around `TypeVarScope` is a bit "opaque" and requires a refactoring anyway, so maybe it makes sense to this in a separate PR, up to you.